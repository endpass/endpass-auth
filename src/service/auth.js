// @ts-check
import get from 'lodash/get';
import request from '@/class/singleton/request';
import { AUTH_STATUS_CODE } from '@/constants';

const identityBaseUrl = ENV.VUE_APP_IDENTITY_API_URL;

const TIMEOUT_DEFAULT = 1500;

/**
 * @type {object}
 */
const CODE_TO_STATUS = {
  200: AUTH_STATUS_CODE.LOGGED_IN,
  401: AUTH_STATUS_CODE.NOT_LOGGED,
  403: AUTH_STATUS_CODE.NEED_PERMISSION,
};

/**
 * @param {Promise} promise
 * @return {Promise<any>}
 */
const withSuccess = promise => {
  return promise.then(res => {
    if (res && !res.success) throw new Error(res.message);

    return res;
  });
};

/**
 * @param {Function} handler
 * @return {number}
 */
const createTimeout = handler => setTimeout(handler, TIMEOUT_DEFAULT);

/**
 * @param {string} email
 * @return {Promise<any>}
 */
const getAuthChallenge = email => {
  return withSuccess(
    request.post(`${identityBaseUrl}/auth`, {
      email,
    }),
  );
};

/**
 * @param {string} password
 * @param {string} originHost
 * @return {Promise<any>}
 */
const setAuthPermission = async (password, originHost) => {
  const res = await request.post(`${identityBaseUrl}/auth/permission`, {
    password,
    originHost,
  });
  return res;
};

/**
 * @param {object} params
 * @param {string} params.email
 * @param {string} params.code
 * @param {string} params.password
 * @param {boolean} params.isSignUp
 * @param {boolean} params.isRemember
 * @return {Promise<any>}
 */
const authWithCode = async ({
  email,
  code,
  password,
  isSignUp,
  isRemember,
}) => {
  const url = isSignUp
    ? `${identityBaseUrl}/auth/signup`
    : `${identityBaseUrl}/auth/token`;

  return withSuccess(
    request.post(url, {
      email,
      code,
      password,
      remember: isRemember,
    }),
  );
};

/**
 * @param {string} idToken
 * @return {Promise<any>}
 */
const authWithGoogle = idToken =>
  request
    .post(`${identityBaseUrl}/auth/google`, {
      token: idToken,
    })
    .then(res => {
      if (!res.success) throw new Error(res.message);
      return res;
    })
    .catch(err => {
      throw err.response.data;
    });

/**
 * @param {string} code
 * @return {Promise<any>}
 */
const authWithGitHub = code =>
  request
    .post(`${identityBaseUrl}/auth/github`, {
      code,
    })
    .then(res => {
      if (!res.success) throw new Error(res.message);
      return res;
    })
    .catch(err => {
      throw err.response.data;
    });

const logout = () => request.post(`${identityBaseUrl}/logout`);

/**
 * @return {Promise<{expiresAt: number, status: string, hash: string}>}
 */
const getAuthStatus = async () => {
  try {
    const { expiresAt, hash } = await request.get(
      `${identityBaseUrl}/auth/check`,
    );
    return {
      status: AUTH_STATUS_CODE.LOGGED_IN,
      hash,
      expiresAt,
    };
  } catch (e) {
    const statusCode = get(e, ['response', 'status']);
    const status = CODE_TO_STATUS[statusCode] || AUTH_STATUS_CODE.NOT_LOGGED;

    const hash = get(e, ['response', 'data', 'hash'], '');
    const expiresAt = get(e, ['response', 'data', 'expiresAt'], 0);
    return {
      status,
      hash,
      expiresAt,
    };
  }
};

/**
 * @return {Promise<number>}
 */
const waitLogin = () =>
  new Promise((resolve, reject) => {
    /* eslint-disable-next-line */
    const handler = async function() {
      try {
        const { status } = await getAuthStatus();

        if (
          status === AUTH_STATUS_CODE.LOGGED_IN ||
          status === AUTH_STATUS_CODE.NEED_PERMISSION
        ) {
          return resolve();
        }

        createTimeout(handler);
      } catch (err) {
        return reject(err);
      }
    };

    return handler();
  });

/**
 * @param {string} email
 * @return {Promise<string>}
 */
const getRecoveryIdentifier = email =>
  request
    .get(`${identityBaseUrl}/auth/recover?email=${encodeURIComponent(email)}`)
    .then(res => {
      if (!res.success) throw new Error(res.message);

      return res.message;
    });

/**
 * @param {string} email
 * @return {Promise<number>}
 */
const sendEmailCode = async email => {
  const { timeout } = await request.post(`${identityBaseUrl}/auth/code`, {
    email,
  });

  return timeout;
};

/**
 * Send sms with code for disabling otp
 * @param {string} email
 * @returns {Promise<any>}
 */
const sendOtpRecoverSms = async email => {
  try {
    return withSuccess(
      request.get(
        `${identityBaseUrl}/auth/recover?email=${encodeURIComponent(email)}`,
      ),
    );
  } catch (error) {
    const { response = {} } = error;
    const { status } = response;
    error.code = status;

    throw error;
  }
};

/**
 * Disable otp setting
 * @param {object} param
 * @param {string} param.email
 * @param {string} param.code
 * @returns {Promise<any>}
 */
const disableOtpViaSms = async ({ email, code }) => {
  return withSuccess(
    request.post(`${identityBaseUrl}/auth/recover`, {
      email,
      code,
    }),
  );
};

/**
 * @return {string}
 */
const getSeedTemplateUrl = () => `${identityBaseUrl}/auth/recover/template`;

export default {
  getAuthStatus,
  setAuthPermission,
  getAuthChallenge,
  authWithCode,
  authWithGoogle,
  authWithGitHub,
  sendEmailCode,
  logout,
  waitLogin,
  getRecoveryIdentifier,
  sendOtpRecoverSms,
  disableOtpViaSms,
  getSeedTemplateUrl,
};
