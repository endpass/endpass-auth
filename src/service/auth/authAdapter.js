import { CHALLENGE_TYPES } from '@/constants';

function getChallengeType(challenge) {
  if (challenge === 'sms') {
    return CHALLENGE_TYPES.SMS_OTP;
  }

  if (challenge === 'otp') {
    return CHALLENGE_TYPES.APP_OTP;
  }

  return CHALLENGE_TYPES.EMAIL_OTP;
}

export default function(authData) {
  return {
    ...authData,
    challenge: {
      ...authData.challenge,
      challengeType: getChallengeType(authData.challenge),
    },
  };
}
