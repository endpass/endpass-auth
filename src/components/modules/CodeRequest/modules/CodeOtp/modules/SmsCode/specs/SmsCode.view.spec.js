import { shallowMount, createLocalVue } from '@vue/test-utils';
import { code } from '@unitFixtures/auth';
import VeeValidate from 'vee-validate';
import SmsCodeView from '@/components/modules/CodeRequest/modules/CodeOtp/modules/SmsCode/SmsCode.view';
import setupI18n from '@/locales/i18nSetup';

const localVue = createLocalVue();
const i18n = setupI18n(localVue);

localVue.use(VeeValidate);

describe('SmsCodeView', () => {
  let wrapper;
  const createWrapper = (options, props) =>
    shallowMount(SmsCodeView, {
      provide: {
        theme: 'default',
      },
      localVue,
      propsData: {
        isLoading: false,
        error: '',
        ...props,
      },
      sync: false,
      i18n,
      ...options,
    });

  beforeEach(() => {
    jest.clearAllMocks();

    wrapper = createWrapper();
  });

  describe('render', () => {
    it('should correctly render component', () => {
      expect(wrapper.name()).toBe('SmsCodeView');
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render form', () => {
      expect(wrapper.find('[data-test=sms-code-form]').exists()).toBe(true);
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('props', () => {
    describe('error', () => {
      it('should show error from prop', () => {
        expect.assertions(1);

        const error = 'error';

        wrapper = createWrapper(null, { error });

        expect(wrapper.find('[data-test=code-input]').attributes().error).toBe(
          error,
        );
      });

      it('should remove error when empty', async () => {
        expect.assertions(2);

        const error = 'error';

        wrapper = createWrapper(null, { error });
        await wrapper.vm.$nextTick();

        expect(wrapper.find('[data-test=code-input]').attributes().error).toBe(
          error,
        );

        wrapper.setProps({ error: '' });
        await wrapper.vm.$nextTick();

        expect(
          wrapper.find('[data-test=code-input]').attributes().error,
        ).toBeFalsy();
      });

      it('should update error when not empty', () => {
        expect.assertions(2);

        const error = 'error';
        const newError = 'error';

        wrapper = createWrapper(null, { error });

        expect(wrapper.find('[data-test=code-input]').attributes().error).toBe(
          error,
        );

        wrapper.setProps({ error: newError });

        expect(wrapper.find('[data-test=code-input]').attributes().error).toBe(
          newError,
        );
      });
    });
  });
  describe('behavior', () => {
    it('should emit submit event', async () => {
      expect.assertions(4);

      expect(wrapper.emitted().submit).toBeUndefined();

      wrapper.find('[data-test=code-input]').vm.$emit('input', code);
      await global.flushPromises();

      expect(
        wrapper.find('[data-test=submit-button]').attributes().disabled,
      ).toBeUndefined();

      wrapper.find('[data-test=sms-code-form]').trigger('submit');

      expect(wrapper.emitted().submit.length).toBe(1);
      expect(wrapper.emitted().submit[0]).toEqual([
        { code, isRemember: false },
      ]);
    });

    it('should emit submit event with remember', async () => {
      expect.assertions(4);

      expect(wrapper.emitted().submit).toBeUndefined();

      wrapper.find('[data-test=code-input]').vm.$emit('input', code);
      wrapper.find('[data-test=remember-me-checkbox]').vm.$emit('input', true);
      await global.flushPromises();

      expect(
        wrapper.find('[data-test=submit-button]').attributes().disabled,
      ).toBeUndefined();

      wrapper.find('[data-test=sms-code-form]').trigger('submit');

      expect(wrapper.emitted().submit.length).toBe(1);
      expect(wrapper.emitted().submit[0]).toEqual([{ code, isRemember: true }]);
    });

    it('should emit send code event', () => {
      expect(wrapper.emitted()['send-code']).toBeUndefined();

      wrapper.find('send-code-stub').vm.$emit('send-code');

      expect(wrapper.emitted()['send-code'].length).toBe(1);
      expect(wrapper.emitted()['send-code'][0]).toEqual([]);
    });
  });

  describe('validation', () => {
    describe('code input', () => {
      it('should not show error by default', () => {
        expect(
          wrapper.find('[data-test=code-input]').attributes().error,
        ).toBeFalsy();
      });

      it('should not show error with valid data', async () => {
        expect.assertions(1);

        wrapper.find('[data-test=code-input]').vm.$emit('input', code);
        await global.flushPromises();

        expect(
          wrapper.find('[data-test=code-input]').attributes().error,
        ).toBeFalsy();
      });

      it('should show error with invalid code', async () => {
        expect.assertions(1);

        wrapper.find('[data-test=code-input]').vm.$emit('input', 'short');
        await global.flushPromises();

        expect(
          wrapper.find('[data-test=code-input]').attributes().error,
        ).toMatch('6 digits');
      });

      it('should show error with empty code', async () => {
        expect.assertions(1);

        wrapper.find('[data-test=code-input]').vm.$emit('input', 'short');
        await global.flushPromises();
        wrapper.find('[data-test=code-input]').vm.$emit('input', '');
        await global.flushPromises();

        expect(
          wrapper.find('[data-test=code-input]').attributes().error,
        ).toMatch('required');
      });
    });

    describe('submit button', () => {
      it('should disable submit by default', () => {
        expect(
          wrapper.find('[data-test=submit-button]').attributes().disabled,
        ).toBe('true');
      });

      it('should enable submit with valid code', async () => {
        expect.assertions(1);

        wrapper.find('[data-test=code-input]').vm.$emit('input', code);
        await global.flushPromises();

        expect(
          wrapper.find('[data-test=submit-button]').attributes().disabled,
        ).toBeUndefined();
      });

      it('should disable submit with invalid code', async () => {
        expect.assertions(1);

        wrapper.find('[data-test=code-input]').vm.$emit('input', code);
        await global.flushPromises();

        wrapper.find('[data-test=code-input]').vm.$emit('input', 'short');
        await global.flushPromises();

        expect(
          wrapper.find('[data-test=submit-button]').attributes().disabled,
        ).toBe('true');
      });
    });
  });
});
