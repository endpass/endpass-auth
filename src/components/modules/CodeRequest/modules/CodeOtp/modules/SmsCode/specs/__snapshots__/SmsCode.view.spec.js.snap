// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmsCodeView render should correctly render component 1`] = `
<form data-test="sms-code-form">
  <v-title-stub>
    Verify SMS code
  </v-title-stub>
  <v-description-stub>
    Enter the 6 digit code sent on your phone
  </v-description-stub>
  <form-item-stub>
    <v-input-stub label="" tooltiplabel="" value="" error="" description="" skin="primary" data-vv-as="code" data-vv-name="code" name="code" placeholder="Enter received code..." data-test="code-input"></v-input-stub>
  </form-item-stub>
  <form-item-stub class="v-mb-24">
    <v-checkbox-remember-stub data-test="remember-me-checkbox"></v-checkbox-remember-stub>
  </form-item-stub>
  <form-item-stub class="v-mb-24">
    <v-button-stub skin="primary" size="normal" type="submit" data-test="submit-button" disabled="true">
      Confirm
    </v-button-stub>
  </form-item-stub>
  <form-row-stub class="v-fs-14 v-text-center">
    <send-code-stub timeout="60000"></send-code-stub>
  </form-row-stub>
</form>
`;

exports[`SmsCodeView render should render form 1`] = `
<form data-test="sms-code-form">
  <v-title-stub>
    Verify SMS code
  </v-title-stub>
  <v-description-stub>
    Enter the 6 digit code sent on your phone
  </v-description-stub>
  <form-item-stub>
    <v-input-stub label="" tooltiplabel="" value="" error="" description="" skin="primary" data-vv-as="code" data-vv-name="code" name="code" placeholder="Enter received code..." data-test="code-input"></v-input-stub>
  </form-item-stub>
  <form-item-stub class="v-mb-24">
    <v-checkbox-remember-stub data-test="remember-me-checkbox"></v-checkbox-remember-stub>
  </form-item-stub>
  <form-item-stub class="v-mb-24">
    <v-button-stub skin="primary" size="normal" type="submit" data-test="submit-button" disabled="true">
      Confirm
    </v-button-stub>
  </form-item-stub>
  <form-row-stub class="v-fs-14 v-text-center">
    <send-code-stub timeout="60000"></send-code-stub>
  </form-row-stub>
</form>
`;
